File: matlab/R2RILS.m
function [X_hat, U_hat, lambda_hat, V_hat, observed_RMSE, iter, convergence_flag] = ...
    R2RILS(X,omega,r,opts)
% function [X_hat, U_hat, lambda_hat, V_hat, observed_RMSE, iter, convergence_flag] = R2RILS_v2(X,omega,r,opts)
%
% WRITTEN BY BAUCH, NADLER & ZILBER / 2020
%
% INPUT: 
% X = matrix with observed entries in the set omega
% omega = list of size nv * 2 of pairs (i,j) ; nv=total number of visible entries
% r = target rank of reconstructed matrix
% opts: options meta-variable (see opts_default for details)
%

% default values for option variables
opts_default.verbose = 0;                       % display intermediate results
opts_default.max_iter = 50;                     % total number of iterations
opts_default.LSQR_col_norm = 0;                 % normalize columns of least squares matrix
opts_default.LSQR_max_iter = 4000;              % maximal number of iterations for the LSQR solver
opts_default.LSQR_tol = 1e-15;                  % tolerance of LSQR solver
opts_default.LSQR_smart_tol = 1;                % use LSQR_tol==observed_RMSE^2 when obersved_RMSE is low enough
opts_default.LSQR_smart_obj_min = 1e-5;         % observed_RMSE threshold to start using LSQR smart tol
opts_default.init_option = 0;                   % 0 for SVD, 1 for random, 2 for opts.init_U, opts.init_V
opts_default.init_U = NaN;                      % if opts.init_option==2, use this initialization for U
opts_default.init_V = NaN;                      % if opts.init_option==2, use this initialization for V
opts_default.weight_previous_estimate = 1.0;    % different averaging weight for previous estimate (should prevent oscillations)
opts_default.weight_from_iter = 40;             % start different averaging from itertion
opts_default.weight_every_iter = 7;             % use different averaging when mod(iter, weight_every_iter) < 2
opts_default.early_stopping_RMSE_abs = 5e-14;   % small observed_RMSE threshold (relevant to noise-free case). -1 for disabled
opts_default.early_stopping_rel = -1;           % small relative X_hat difference thrshold. -1 for disabled
opts_default.early_stopping_RMSE_rel = -1;      % small relative observed_RMSE difference. -1 for disabled

% for each unset option set its default value
fn = fieldnames(opts_default);
for k=1:numel(fn)
    if ~isfield(opts,fn{k}) || isempty(opts.(fn{k}))
        opts.(fn{k}) = opts_default.(fn{k});
        fn{k} = opts_default.(fn{k});
    end
end

% some definitions
[nr, nc] = size(X);   %nr,nc = number of rows / colums
m = (nc + nr) * r;  % total number of variables in single update iteration
nv = size(omega,1);   % number of observed entries
rhs = zeros(nv,1);   % vector of visible entries in matrix X
for counter=1:nv
    rhs(counter) = X(omega(counter,1),omega(counter,2)); 
end

% initialize U and V (of sizes nr x r and nc x r)
if opts.init_option == 0
    % initialization by rank-r SVD of observed matrix
    [U, ~, V] = svds(X,r);
elseif opts.init_option == 1
    % initialization by random orthogonal matrices
    Z = randn(nr,r);
    [U, ~, ~] = svd(Z,'econ'); 
    Z = randn(nc,r);
    [V, ~, ~] = svd(Z,'econ'); 
else
    % initiazliation by user-defined matrices
    U = opts.U_init;
    V = opts.V_init; 
end

X_hat_previous = zeros(nr,nc);   % previous intermediate rank 2r estimate
X_r_previous = zeros(nr,nc);   % rank-r projection of previous intermediate rank 2r estimate
observed_RMSE = zeros(opts.max_iter,1);   % of the rank-r projection of intermediate rank 2r matrix
X_max = max(max(abs(X)));
current_observed_RMSE = X_max;
best_RMSE = X_max; % initialized with max absolute value of all observed entries
convergence_flag = 0;

for loop_idx = 1:opts.max_iter
    
    if opts.verbose && (mod(loop_idx,5)==0)
        fprintf('INSIDE R2RILS-V2 loop_idx  %d/%d\n',loop_idx,opts.max_iter);
    end

    % Z^T = [a(coordinate 1)  a(coordinate 2) ... a(coordinate nc) | b(coordinate 1) ... b(coordinate nr) ]
    % Z^T = [ Vnew and then Unew ]
    Z = zeros(m,1);    % Z is a long vector with a vectorization of A and of B (notation in paper)
    A = zeros(nv,m);    % matrix of least squares problem
    B = zeros(1,m);    % used to normalize A columns if needed

    % contsruction of A and B
    for counter=1:nv
        j = omega(counter,1);
        k = omega(counter,2);
        index=r*(k-1)+1;
        A(counter,index:index+r-1) = U(j,:);
        B(index:index+r-1) = B(index:index+r-1) + U(j,:).^2;
        index = r*nc + r*(j-1)+1; 
        A(counter,index:index+r-1) = V(k,:);
        B(index:index+r-1) = B(index:index+r-1) + V(k,:).^2;
    end
    
    B = sqrt(B);
    if opts.LSQR_col_norm
        for j=1:size(A,2)
            A(:,j)=A(:,j)/B(j);
        end
    end
    A = sparse(A); 
    
    % determine tolerance for LSQR solver
    LSQR_tol = opts.LSQR_tol;
    if opts.LSQR_smart_tol
        LSQR_tol = min(opts.LSQR_smart_obj_min, current_observed_RMSE^2);
    end
    LSQR_tol = max(LSQR_tol, 2*eps);    % to supress warning
    % solve the least squares problem
    [Z, flag, relres_Z, iter_Z] = lsqr(A,rhs,LSQR_tol,opts.LSQR_max_iter);
        % LSQR finds the minimum norm solution and is much faster than lsqminnorm
    %disp([flag relres_Z iter_Z]);
    if opts.LSQR_col_norm
        Z = Z./B';
    end
    
    % construct Utilde and Vtilde from the entries of the long vector Z 
    Utilde = zeros(size(U)); Vtilde = zeros(size(V)); 
    nc_list = r * [0:1:(nc-1)]; 
    for i=1:r
        Vtilde(:,i) = Z(i+nc_list); 
    end
    nr_list = r * [0:1:(nr-1)]; 
    start_idx = r*nc; 
    for i=1:r
        Utilde(:,i) = Z(start_idx + i + nr_list);
    end
   
    % intermediate rank-2r estimate
    X_hat = U * Vtilde' + Utilde * V';
    
    % rank r projection of intermediate rank-2r solution
    [U_r, lambda_r, V_r] = svds(X_hat,r); 
    X_r = U_r * lambda_r * V_r'; 
    
    % calculate RMSE and update X_out if needed
    current_observed_RMSE = sqrt(sum(sum( (abs(X)>0).*(X_r - X).^2 )) / nv);
    observed_RMSE(loop_idx) = current_observed_RMSE; 
    if current_observed_RMSE < best_RMSE
        best_RMSE = current_observed_RMSE; 
        X_out = X_r; 
    end
    
    % col-normalize Utilde and Vtilde
    normU = sqrt(sum(Utilde.^2)); 
    Utilde = Utilde * diag(1./normU);  
    normV = sqrt(sum(Vtilde.^2)); 
    Vtilde = Vtilde * diag(1./normV);  

    % calculate next estimate of U, V by averaging previous estimate with
    % Utilde, Vtilde
    weight = 1.0; 
    if (loop_idx > opts.weight_from_iter && mod(loop_idx,opts.weight_every_iter)<2)         
        weight = opts.weight_previous_estimate;
        if opts.verbose
            fprintf('INSIDE R2RILS-V2 using different weight for previous estimate\n');
        end
    end
    U = (weight*U + Utilde);
    V = (weight*V + Vtilde); 
    
    % col-normalize U and V
    normU = sqrt(sum(U.^2)); 
    U = U * diag(1./normU);
    normV = sqrt(sum(V.^2)); 
    V = V * diag(1./normV); 

    % calculate new RMSE
    X_hat_diff = sqrt(sum(sum((X_hat-X_hat_previous).^2)))/sqrt(nr*nc);
    
    if opts.verbose
        fprintf('INSIDE R2RILS-V2 loop_idx %3d \t DIFF X_hat %6d DIFF X_r %6d observed_RMSE %6d\n',...
            loop_idx,X_hat_diff,...
            sqrt(sum(sum((X_r-X_r_previous).^2)))/sqrt(nr*nc), current_observed_RMSE);
        if loop_idx>1 
            fprintf('INSIDE R2RILS-V2 \t\t\t\t oberved_RMSE(new) / observed_RMSE(old)-1: %f\n',current_observed_RMSE/observed_RMSE(loop_idx-1)-1); 
        end
    end
    
    % early stopping criteria
    if current_observed_RMSE < opts.early_stopping_RMSE_abs
        if opts.verbose
            fprintf('INSIDE R2RILS-V2 Early stopping, small error on observed entries\n'); 
        end
        convergence_flag = 1;
    elseif X_hat_diff < opts.early_stopping_rel
        if opts.verbose
            fprintf('INSIDE R2RILS-V2 Early stopping, X_hat does not change\n'); 
        end
        convergence_flag = 1;
    elseif loop_idx > 1 && ...
            abs(observed_RMSE(loop_idx-1)/current_observed_RMSE-1) < opts.early_stopping_RMSE_rel
        if opts.verbose
            fprintf('INSIDE R2RILS-V2 Early stopping, observed_RMSE does not change\n'); 
        end
        convergence_flag = 1;
    end
    if convergence_flag
        break
    end
    
    X_hat_previous = X_hat;
    X_r_previous = X_r;
end

% return rank-r SVD of X_out, the rank-r matrix with the lowest observed
% RMSE during the iterations
[U_hat, lambda_hat, V_hat] = svds(X_out,r);
%X_hat = U_hat * lambda_hat * V_hat';
X_hat = X_out;
iter = loop_idx;

File: python/R2RILS.py
"""
WRITTEN BY BAUCH, NADLER & ZILBER / 2020
"""


import numpy as np
from scipy import sparse
from scipy.sparse import linalg
from sklearn.preprocessing import normalize

# initialization options
INIT_WITH_SVD = 0
INIT_WITH_RANDOM = 1
INIT_WITH_USER_DEFINED = 2

def R2RILS(X, omega, rank, verbose=True, max_iter=100,
           lsqr_col_norm=False, lsqr_max_iter=1000, lsqr_tol=1e-15,
           lsqr_smart_tol=True, lsqr_smart_obj_min=1e-5,
           init_option=INIT_WITH_SVD , init_U=None, init_V=None,
           weight_previous_estimate=1.0, weight_from_iter=40, weight_every_iter=7,
           early_stopping_rmse_abs=5e-14, early_stopping_rel=-1, early_stopping_rmse_rel=-1):
    """
    Run R2RILS algorithm.
    :param ndarray X: Input matrix (m,n). Unobserved entries should be zero.
    :param ndarray omega: Mask matrix (m,n). 1 on observed entries, 0 on unobserved.
    :param int rank: Underlying rank matrix.
    :param bool verbose: if True, display intermediate results.
    :param int max_iter: Maximal number of iterations to perform.
    :param bool lsqr_col_norm: if True, columns of the LSQR matrix are normalized.
    :param int lsqr_max_iter: max number of iterations fot the LSQR solver.
    :param float lsqr_tol: tolerance of LSQR solver.
    :param bool lsqr_smart_tol: if True, when objective <= lsqr_smart_obj_min, use lsqr_tol=objective**2.
    :param float lsqr_smart_obj_min: maximal obejctive to begin smart tolerance from.
    :param init_option: how to initialize U and V: 0 for SVD, 1 for random, 2 for user-defined.
    :param ndarray init_U: U initialization (m,rank), used in case init_option==2.
    :param ndarray init_V: V initialization (n,rank), used in case init_option==2.
    :param float weight_previous_estimate: different averaging weight for the previous estimate of U, V.
    :param int weight_from_iter: start the different weighting from this iteration number.
    :param int weight_every_iter: use the different weighting when iter_num % weight_every_iter < 2.
    :param float early_stopping_rmse_abs: eps for absolute difference between X_hat and X (RMSE), -1 for disabled.
    :param float early_stopping_rel: eps for relative difference of X_hat between iterations, -1 for disabled.
    :param float early_stopping_rmse_rel: eps for relative difference of RMSE between iterations, -1 for disabled.
    :return: R2RILS's estimate, and convergence flag (True if early steopped, False if iterations exceeded max).
    """
    m, n = X.shape
    num_visible_entries = np.count_nonzero(omega)

    # initial estimate
    if init_option == INIT_WITH_SVD:
      (U, _, V) = linalg.svds(X, k=rank, tol=1e-16)
      V = V.T
    elif init_option == INIT_WITH_RANDOM:
      U = np.random.randn(m, rank)
      V = np.random.randn(n, rank)
      U = np.linalg.qr(U)[0]
      V = np.linalg.qr(V)[0]
    else:
      U = init_U
      V = init_V
    # change the shapes from (m,r) and (n,r) to (r,m) and (r,n)
    U, V = U.T, V.T

    # generate sparse indices to accelerate future operations.
    sparse_matrix_rows, sparse_matrix_columns = generate_sparse_matrix_entries(omega, rank, m, n)
    # generate (constant) b for the least squares problem
    b = generate_b(X, omega, m, n)

    # start iterations
    early_stopping_flag = False
    X_hat_previous = np.dot(U.T, V)
    objective_previous = 1.
    best_RMSE = np.max(np.abs(X))
    X_hat_final = X_hat_previous  # X_hat_final will store the estimation with the best RMSE
    iter_num = 0
    objective = 1
    while iter_num < max_iter and not early_stopping_flag:
        if verbose and (iter_num % 5 == 0):
          print("iteration {}/{}".format(iter_num, max_iter))
        iter_num += 1
        
        # determine LSQR tolerance
        tol = lsqr_tol
        if lsqr_smart_tol:
          tol = min(lsqr_smart_obj_min, objective**2)
        
        # solve the least squares problem
        A = generate_sparse_A(U, V, omega, sparse_matrix_rows, sparse_matrix_columns, num_visible_entries, m, n,
                              rank)
        scale_vec = np.ones((A.shape[1]),)
        if lsqr_col_norm:
          A, scale_vec = rescale_A(A)
        x = linalg.lsqr(A, b, atol=tol, btol=tol, iter_lim=lsqr_max_iter)[0]
        x = x * scale_vec
        x = convert_x_representation(x, rank, m, n)
        
        # get new estimate for X_hat (project 2r onto r) and calculate its objective
        X_hat = get_estimated_value(x, U, V, rank, m, n)
        (U_r, Sigma_r, V_r) = linalg.svds(X_hat, k=rank, tol=1e-16)
        X_hat = U_r @ np.diag(Sigma_r) @ V_r
        objective = np.linalg.norm((X_hat - X) * omega, ord='fro') / np.sqrt(num_visible_entries)
        if verbose:
          print("objective: {}".format(objective))
        if objective < best_RMSE:
          X_hat_final = X_hat
          best_RMSE = objective

        # obtain new estimates for U and V
        U_tilde, V_tilde = get_U_V_from_solution(x, rank, m, n)
        # ColNorm U_tilde, V_tilde
        U_tilde = normalize(U_tilde, axis=1)
        V_tilde = normalize(V_tilde, axis=1)
        # average with previous estimate
        weight = 1.
        if (iter_num >= weight_from_iter) and (iter_num % weight_every_iter < 2):
          # this should prevent oscillations
          weight = weight_previous_estimate
          if verbose:
            print("using different weighting for previous estimate")
        U = normalize(weight * U + U_tilde, axis=1)
        V = normalize(weight * V + V_tilde, axis=1)

        # check early stopping criteria
        early_stopping_flag = False
        if early_stopping_rmse_abs > 0:
          early_stopping_flag |= objective < early_stopping_rmse_abs
        if early_stopping_rel > 0:
          early_stopping_flag |= np.linalg.norm(X_hat - X_hat_previous,
                                                ord='fro') / np.sqrt(m * n) < early_stopping_rel
        if early_stopping_rmse_rel > 0:
          early_stopping_flag |= np.abs(objective / objective_previous - 1) < early_stopping_rmse_rel

        # update previous X_hat and objective
        X_hat_previous = X_hat
        objective_previous = objective

    # return
    convergence_flag = iter_num < max_iter
    return X_hat_final, convergence_flag


def rescale_A(A):
    A = sparse.csc_matrix(A)
    scale_vec = 1. / linalg.norm(A, axis=0)
    return normalize(A, axis=0), scale_vec


def convert_x_representation(x, rank, m, n):
    recovered_x = np.array([x[k * rank + i] for i in range(rank) for k in range(n)])
    recovered_y = np.array([x[rank * n + j * rank + i] for i in range(rank) for j in range(m)])
    return np.append(recovered_x, recovered_y)


def generate_sparse_matrix_entries(omega, rank, m, n):
    row_entries = []
    columns_entries = []
    row = 0
    for j in range(m):
        for k in range(n):
            if 0 != omega[j][k]:
                # add indices for U entries
                for l in range(rank):
                    columns_entries.append(k * rank + l)
                    row_entries.append(row)
                # add indices for V entries
                for l in range(rank):
                    columns_entries.append((n + j) * rank + l)
                    row_entries.append(row)
                row += 1
    return row_entries, columns_entries


def generate_sparse_A(U, V, omega, row_entries, columns_entries, num_visible_entries, m, n, rank):
    U_matrix = np.array(U).T
    V_matrix = np.array(V).T
    # we're generating row by row
    data_vector = np.concatenate(
        [np.concatenate([U_matrix[j], V_matrix[k]]) for j in range(m) for k in range(n) if 0 != omega[j][k]])
    return sparse.csr_matrix(sparse.coo_matrix((data_vector, (row_entries, columns_entries)),
                                               shape=(num_visible_entries, rank * (m + n))))

def generate_b(X, omega, m, n):
    return np.array([X[j][k] for j in range(m)
                     for k in range(n) if 0 != omega[j][k]])


def get_U_V_from_solution(x, rank, m, n):
    V = np.array([x[i * n:(i + 1) * n] for i in range(rank)])
    U = np.array([x[rank * n + i * m: rank * n + (i + 1) * m] for i in range(rank)])
    return U, V


def get_estimated_value(x, U, V, rank, m, n):
    # calculate U's contribution
    estimate = np.sum(
        [np.dot(U[i].reshape(m, 1), np.array(x[i * n:(i + 1) * n]).reshape(1, n)) for i in
         range(rank)],
        axis=0)
    # calculate V's contribution
    estimate += np.sum(
        [np.dot(x[rank * n + i * m: rank * n + (i + 1) * m].reshape(m, 1),
                V[i].reshape(1, n))
         for i in range(rank)], axis=0)
    return estimate
